// @apps-in-toss/native-modules/src/MiniAppModule/native-modules/getDeclaredAgeRange.ts
import { Platform } from 'react-native';
import { isMinVersionSupported } from './isMinVersionSupported';
import { MiniAppModule } from '../../natives';

export interface DeclaredAgeRange {
  status: 'SHARING' | 'DECLINED_SHARING' | 'UNSUPPORTED_OS' | 'NOT_AVAILABLE' | 'INVALID_REQUEST' | 'FAILED';
  lowerBound?: number;
  upperBound?: number;
  ageRangeDeclaration?: string;
  activeParentalControls: string[];
  isEligibleForAgeFeatures?: boolean;
  errorCode?: string;
}

const isSupportedPlatform = Platform.OS === 'ios' && parseInt(Platform.Version, 10) >= 26;
const MIN_IOS_TOSSAPP_VERSION = '5.266.0';
const isSupportedTossApp = isMinVersionSupported({ ios: MIN_IOS_TOSSAPP_VERSION, android: 'never' });

export async function getDeclaredAgeRange(params: { ageGates: number[] }): Promise<DeclaredAgeRange> {
  if (!isSupportedPlatform) {
    throw new Error('getDeclaredAgeRange는 iOS 26 이상에서만 사용 가능합니다.');
  }

  if (!isSupportedTossApp) {
    throw new Error(`getDeclaredAgeRange는 토스앱 ${MIN_IOS_TOSSAPP_VERSION} 버전 이상에서만 사용 가능합니다.`);
  }

  return await MiniAppModule.postMessage('getDeclaredAgeRange', params, {});
}
getDeclaredAgeRange.isSupported = () => isSupportedPlatform && isSupportedTossApp;
